### PR TITLE
camkes: remove CMAKE_INTERFACE_INCLUDES

### PR DESCRIPTION
**Pull request by axel-h**
Remove generated shell script camkes.h.cmd and camkes.h.in. Use a
lightweight camkes.h instead that includes a component-specific
generated header camkes-component-<component>.h and the interface
template headers.

### CODE REVIEW

**Comment by axel-h** [comment]
@ssrg-bamboo `test`

**Comment by ssrg-bamboo** [comment]
Hello, I'm a bot! I'll bring this PR into Trustworthy Systems and run some tests

**Comment by ssrg-bamboo** [comment]
## Test results for commit cea0c40ae8e0832541239f5225a50f5855d73bbe

### Summary

| Test name | Result |
| --- | --- |
| CAmkES - CAmkES Unit Tests - EXT-axel-h-patch-axel-9-49 | FAILED |
| Code Quality - Check licenses - camkes-tool - EXT-axel-h-patch-axel-9-49 | SUCCESSFUL |
| Code Quality - Style - camkes-tool - EXT-axel-h-patch-axel-9-49 | SUCCESSFUL |

### Detailed failure logs

| Project | Stage | Result summary |
| --- | --- | --- |
| CAmkES - CAmkES Unit Tests - EXT-axel-h-patch-axel-9-49 | Build | 1 of 1 job failed |

**Job: Run tests**
Build logs

    06-Dec-2020 10:02:39	Build CAmkES - CAmkES Unit Tests - EXT-axel-h-patch-axel-9-49 - Run tests #1 (CAMKES-CUT188-JOB-1) started building on agent A local agent 1, bamboo version: 7.1.4
    06-Dec-2020 10:02:39	I run on the Bamboo server!
    06-Dec-2020 10:02:39	Build working directory is /var/atlassian/application-data/bamboo/xml-data/build-dir/140083221/CAMKES-CUT188-JOB
    06-Dec-2020 10:02:39	Executing build CAmkES - CAmkES Unit Tests - EXT-axel-h-patch-axel-9-49 - Run tests #1 (CAMKES-CUT188-JOB-1)
    06-Dec-2020 10:02:39	Preparing artifact 'code' for use at /var/atlassian/application-data/bamboo/xml-data/build-dir/140083221/CAMKES-CUT188-JOB/. (location: .)
    06-Dec-2020 10:02:39	Artifact 'code' prepared in 11.53 ms at /var/atlassian/application-data/bamboo/xml-data/build-dir/140083221/CAMKES-CUT188-JOB/./code.tar.gz
    06-Dec-2020 10:02:39	Running pre-build action: VCS Version Collector
    06-Dec-2020 10:02:39	Starting task 'Untar code' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    06-Dec-2020 10:02:39	+ tar xf code.tar.gz --recursive-unlink
    06-Dec-2020 10:02:39	Finished task 'Untar code' with result: Success
    06-Dec-2020 10:02:39	Starting task 'Set permissions so Bamboo can cleanup' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    06-Dec-2020 10:02:39	+ chmod -R o+rwx /var/atlassian/application-data/bamboo/xml-data/build-dir/140083221/CAMKES-CUT188-JOB
    06-Dec-2020 10:02:39	Finished task 'Set permissions so Bamboo can cleanup' with result: Success
    06-Dec-2020 10:02:39	Starting task 'Install dependencies' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    06-Dec-2020 10:02:39	+ pip2 install --user --upgrade camkes-deps
    06-Dec-2020 10:02:39	/var/atlassian/application-data/bamboo/temp/CAMKES-CUT188-JOB-1-ScriptBuildTask-5711000774807173656.sh: line 6: pip2: command not found
    06-Dec-2020 10:02:39	Failing task since return code of [/var/atlassian/application-data/bamboo/temp/CAMKES-CUT188-JOB-1-ScriptBuildTask-5711000774807173656.sh] was 127 while expected 0
    06-Dec-2020 10:02:39	Finished task 'Install dependencies' with result: Failed
    06-Dec-2020 10:02:39	Starting task 'Set permissions so Bamboo can cleanup' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    06-Dec-2020 10:02:39	+ chmod -R o+rwx /var/atlassian/application-data/bamboo/xml-data/build-dir/140083221/CAMKES-CUT188-JOB
    06-Dec-2020 10:02:39	Finished task 'Set permissions so Bamboo can cleanup' with result: Success
    06-Dec-2020 10:02:39	Running post build plugin 'NCover Results Collector'
    06-Dec-2020 10:02:39	Running post build plugin 'Artifact Copier'
    06-Dec-2020 10:02:39	Running post build plugin 'npm Cache Cleanup'
    06-Dec-2020 10:02:39	Running post build plugin 'Clover Results Collector'
    06-Dec-2020 10:02:39	Running post build plugin 'Docker Container Cleanup'
    06-Dec-2020 10:02:39	Successfully removed working directory at '/var/atlassian/application-data/bamboo/xml-data/build-dir/140083221/CAMKES-CUT188-JOB'
    06-Dec-2020 10:02:39	Finalising the build...
    06-Dec-2020 10:02:39	Stopping timer.
    06-Dec-2020 10:02:39	Build CAMKES-CUT188-JOB-1 completed.
    06-Dec-2020 10:02:39	Running on server: post build plugin 'NCover Results Collector'
    06-Dec-2020 10:02:39	Running on server: post build plugin 'Build Hanging Detection Configuration'
    06-Dec-2020 10:02:39	Running on server: post build plugin 'Clover Delta Calculator'
    06-Dec-2020 10:02:39	Running on server: post build plugin 'Maven Dependencies Postprocessor'
    06-Dec-2020 10:02:39	All post build plugins have finished
    06-Dec-2020 10:02:39	Generating build results summary...
    06-Dec-2020 10:02:39	Saving build results to disk...
    06-Dec-2020 10:02:39	Store variable context...
    06-Dec-2020 10:02:39	Indexing build results...
    06-Dec-2020 10:02:39	Finished building CAMKES-CUT188-JOB-1.


This is the most I can report on right now, sorry!

**Comment by axel-h** [comment]
seems bamboo is lacking some fundamentals tools: `pip2: command not found`

**Comment by lsf37** [comment]
@LukeMondy could you please have a look at `python2` for bamboo?

**Comment by xurtis** [comment]
@ssrg-bamboo `test`

**Comment by ssrg-bamboo** [comment]
Hello, I'm a bot! I'll bring this PR into Trustworthy Systems and run some tests

**Comment by axel-h** [comment]
@ssrg-bamboo `test`

**Comment by ssrg-bamboo** [comment]
Hello, I'm a bot! I'll bring this PR into Trustworthy Systems and run some tests

**Comment by ssrg-bamboo** [comment]
All the tests we ran have passed! Nice job!

**Review by kent-mcleod — APPROVED** [review]
Thanks for seeing a cleaner way to do this without having `@CMAKE_INTERFACE_INCLUDES@`.

File: camkes/templates/camkes-gen.cmake [hunk 1]
```diff
@@ -295,14 +300,24 @@ RequireFile(CONFIGURE_FILE_SCRIPT configure_file.cmake PATHS ${CMAKE_MODULE_PATH
...
+    set(camkes_h_content "#pragma once\n\n#include \"${camkes_component_h}\"\n")
+    foreach(item IN LISTS CAMKES_INTERFACE_INCLUDES)
+        STRING(APPEND camkes_h_content "#include \"${item}\"\n")
+    endforeach()
+    file(WRITE "${generated_dir}/include/camkes.h" "${camkes_h_content}")
+
+    # It's unclear if for camkes.c we really have to explicitly define a
+    # dependency on camkes.h and ${camkes_component_h}. Without this, the file
+    # ninja.rules does not list any explicit dependency for camkes.c, but
+    # .ninja_deps seems to contain a reference.
```
[the remark this answers is not in the document]
Comment (axel-h) [reply camkes/templates/camkes-gen.cmake:314]: I will remove this now

**Comment by axel-h** [comment]
@ssrg-bamboo `test`

**Comment by ssrg-bamboo** [comment]
Hello, I'm a bot! I'll bring this PR into Trustworthy Systems and run some tests

**Comment by ssrg-bamboo** [comment]
All the tests we ran have passed! Nice job!

**Review by oliver-wm — APPROVED** [review]

**Comment by oliver-wm** [comment]
@ssrg-bamboo `test`

**Comment by ssrg-bamboo** [comment]
Hello, I'm a bot! I'll bring this PR into Trustworthy Systems and run some tests

**Comment by axel-h** [comment]
Can we merge this?

**Comment by axel-h** [comment]
Could you merge this? Besides avoiding the creation or intermediate files, this also reduces the amount of steps ninja executes on rebuilds by one.

File: camkes/templates/camkes-gen.cmake [hunk 2]
```diff
@@ -295,14 +300,24 @@ RequireFile(CONFIGURE_FILE_SCRIPT configure_file.cmake PATHS ${CMAKE_MODULE_PATH
...
     /*- endif -*/
 
-    # Add interface header files to camkes.h
-    file(WRITE "${generated_dir}/include/camkes.h.cmd" "${CMAKE_COMMAND} -DCONFIGURE_INPUT_FILE=\"${generated_dir}/include/camkes.h.in\" \
-        -DCONFIGURE_OUTPUT_FILE=\"${generated_dir}/include/camkes.h\" -DCMAKE_INTERFACE_INCLUDES=\"${CMAKE_INTERFACE_INCLUDES}\" -P ${CONFIGURE_FILE_SCRIPT}")
-    add_custom_command(OUTPUT ${generated_dir}/include/camkes.h
-        COMMAND sh "${generated_dir}/include/camkes.h.cmd"
-        DEPENDS "${generated_dir}/include/camkes.h.in" "${generated_dir}/include/camkes.h.cmd"
-        )
-    set_property(SOURCE /*? i.name ?*//camkes.c PROPERTY OBJECT_DEPENDS ${generated_dir}/include/camkes.h)
```
[the remark this answers is not in the document]
Comment (kent-mcleod) [reply camkes/templates/camkes-gen.cmake:299]: I missed this comment originally.  But it does seem safe to remove these lines when it is a generated header file that is in the same project scope.

**Pull request merged**